src/libraries/Common/src/Roslyn/SyntaxValueProvider_ForAttributeWithMetadataName.cs

```diff
@@ -0,0 +1,188 @@
+    /// cref="System.CLSCompliantAttribute"/>.
+    /// </summary>
+    /// <param name="predicate">A function that determines if the given <see cref="SyntaxNode"/> attribute target (<see
+    /// cref="GeneratorAttributeSyntaxContext.TargetNode"/>) should be transformed.  Nodes that do not pass this
+    /// predicate will not have their attributes looked at at all.</param>
+    /// <param name="transform">A function that performs the transform. This will only be passed nodes that return <see
+    /// langword="true"/> for <paramref name="predicate"/> and which have a matching <see cref="AttributeData"/> whose
+    /// <see cref="AttributeData.AttributeClass"/> has the same fully qualified, metadata name as <paramref
+    /// name="fullyQualifiedMetadataName"/>.</param>
+    public static IncrementalValuesProvider<T> ForAttributeWithMetadataName<T>(
+        this SyntaxValueProvider @this,
+        IncrementalGeneratorInitializationContext context,
+        string fullyQualifiedMetadataName,
+        Func<SyntaxNode, CancellationToken, bool> predicate,
+        Func<GeneratorAttributeSyntaxContext, CancellationToken, T> transform)
+    {
+        var metadataName = fullyQualifiedMetadataName.Contains('+')
+            ? MetadataTypeName.FromFullName(fullyQualifiedMetadataName.Split(s_nestedTypeNameSeparators).Last())
+            : MetadataTypeName.FromFullName(fullyQualifiedMetadataName);
+
+        var nodesWithAttributesMatchingSimpleName = @this.ForAttributeWithSimpleName(metadataName.UnmangledTypeName, predicate);
+
+        var collectedNodes = nodesWithAttributesMatchingSimpleName
+            .Collect()
+            .WithComparer(ImmutableArrayValueComparer<SyntaxNode>.Instance)
+            /*.WithTrackingName("collectedNodes_ForAttributeWithMetadataName")*/;
+
+        // Group all the nodes by syntax tree, so we can process a whole syntax tree at a time.  This will let us make
+        // the required semantic model for it once, instead of potentially many times (in the rare, but possible case of
+        // a single file with a ton of matching nodes in it).
+        var groupedNodes = collectedNodes.SelectMany(
+            static (array, cancellationToken) =>
+                array.GroupBy(static n => n.SyntaxTree)
+                     .Select(static g => new SyntaxNodeGrouping<SyntaxNode>(g)))/*.WithTrackingName("groupedNodes_ForAttributeWithMetadataName")*/;
+
+        var compilationAndGroupedNodesProvider = groupedNodes
+            .Combine(context.CompilationProvider)
+            /*.WithTrackingName("compilationAndGroupedNodes_ForAttributeWithMetadataName")*/;
+
+        var syntaxHelper = CSharpSyntaxHelper.Instance;
+        var finalProvider = compilationAndGroupedNodesProvider.SelectMany((tuple, cancellationToken) =>
+        {
+            var (grouping, compilation) = tuple;
+
+            var result = new ValueListBuilder<T>(Span<T>.Empty);
```

Comment:
this depends on a change @stephentoub says he will make to have ValueListBuilder<T> support Span<T>.EMpty.

Reply:
https://github.com/dotnet/runtime/pull/70917